src/components/PaginationV2/PaginationV2.js

```diff
@@ -0,0 +1,228 @@
+import PropTypes from 'prop-types';
+import React, { Component } from 'react';
+import classnames from 'classnames';
+import Icon from '../Icon';
+import Select from '../Select';
+import SelectItem from '../SelectItem';
+import { equals } from '../../tools/array';
+
+export default class PaginationV2 extends Component {
+  static propTypes = {
+    backwardText: PropTypes.string,
+    className: PropTypes.string,
+    itemRangeText: PropTypes.func,
+    forwardText: PropTypes.string,
+    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
+    itemsPerPageText: PropTypes.string,
+    itemText: PropTypes.func,
+    onChange: PropTypes.func,
+    pageNumberText: PropTypes.string,
+    pageRangeText: PropTypes.func,
+    pageText: PropTypes.func,
+    pageSizes: PropTypes.arrayOf(PropTypes.number).isRequired,
+    totalItems: PropTypes.number,
+    disabled: PropTypes.bool,
+    page: PropTypes.number,
+    pageSize: PropTypes.number,
+    pagesUnknown: PropTypes.bool,
+    isLastPage: PropTypes.bool,
+    pageInputDisabled: PropTypes.bool,
+  };
+
+  static defaultProps = {
+    backwardText: 'Backward',
+    itemRangeText: (min, max, total) => `${min}-${max} of ${total} items`,
+    forwardText: 'Forward',
+    itemsPerPageText: 'Items per page',
+    onChange: () => {},
+    pageNumberText: 'Page Number',
+    pageRangeText: (current, total) => `${current} of ${total} pages`,
+    disabled: false,
+    page: 1,
+    pagesUnknown: false,
+    isLastPage: false,
+    pageInputDisabled: false,
+    itemText: (min, max) => `${min}-${max} items`,
+    pageText: page => `page ${page}`,
+  };
+
+  state = {
+    page: this.props.page,
+    pageSize:
+      this.props.pageSize && this.props.pageSizes.includes(this.props.pageSize)
+        ? this.props.pageSize
+        : this.props.pageSizes[0],
+  };
+
+  componentWillMount() {
+    this.uniqueId = `${Math.floor(Math.random() * 0xffff)}`;
+  }
+
+  componentWillReceiveProps({ pageSizes, page, pageSize }) {
+    if (!equals(pageSizes, this.props.pageSizes)) {
+      this.setState({ pageSize: pageSizes[0], page: 1 });
+    }
+    if (page !== this.props.page) {
+      this.setState({
+        page,
+      });
+    }
+    if (pageSize !== this.props.pageSize) {
+      this.setState({ pageSize });
+    }
+  }
+
+  handleSizeChange = evt => {
+    const pageSize = Number(evt.target.value);
+    this.setState({ pageSize, page: 1 });
+    this.props.onChange({ page: 1, pageSize });
```

Comment:
Should we add this call to the callback of `setState`? So the consumer is guaranteed that `onChange` is called while the component is in the given state.